pkg/apis/metrics/types.go

```diff
@@ -16,14 +16,427 @@ limitations under the License.
+	// Unit: nanocores per second
+	PerCpu []uint64 `json:"perCpu,omitempty"`
+
+	// Time spent in user space.
+	// Unit: nanocores per second
+	User uint64 `json:"user,omitempty"`
+
+	// Time spent in kernel space.
+	// Unit: nanocores per second
+	System uint64 `json:"system,omitempty"`
+}
+
+// CPU usage time statistics.
+type CpuUsage struct {
+	// Total CPU usage.
+	// Units: nanoseconds
+	Total uint64 `json:"total,omitempty"`
+
+	// Per CPU/core usage of the container.
+	// Unit: nanoseconds.
+	PerCpu []uint64 `json:"perCpu,omitempty"`
+
+	// Time spent in user space.
+	// Unit: nanoseconds
+	User uint64 `json:"user,omitempty"`
+
+	// Time spent in kernel space.
+	// Unit: nanoseconds
+	System uint64 `json:"system,omitempty"`
+}
+
+// All CPU usage metrics are cumulative from the creation of the container
+type CpuStats struct {
+	Usage CpuUsage `json:"usage,omitempty"`
+	// Smoothed average of number of runnable threads x 1000.
+	// We multiply by thousand to avoid using floats, but preserving precision.
+	// Load is smoothed over the last 10 seconds. Instantaneous value can be read
+	// from LoadStats.NrRunning.
+	LoadAverage int32 `json:"loadAverage,omitempty"`
+}
+
+type PerDiskStats struct {
+	// Major device number. See https://www.kernel.org/doc/Documentation/devices.txt
+	Major uint64 `json:"major,omitempty"`
+	// Minor device number. See https://www.kernel.org/doc/Documentation/devices.txt
+	Minor uint64            `json:"minor,omitempty"`
+	Stats map[string]uint64 `json:"stats,omitempty"`
+}
+
+// Disk IO stats, as reported by the cgroup block io controller.
+// See https://www.kernel.org/doc/Documentation/cgroups/blkio-controller.txt
+type DiskIoStats struct {
+	IoServiceBytes []PerDiskStats `json:"ioServiceBytes,omitempty"`
+	IoServiced     []PerDiskStats `json:"ioServiced,omitempty"`
+	IoQueued       []PerDiskStats `json:"ioQueued,omitempty"`
+	Sectors        []PerDiskStats `json:"sectors,omitempty"`
+	IoServiceTime  []PerDiskStats `json:"ioServiceTime,omitempty"`
+	IoWaitTime     []PerDiskStats `json:"ioWaitTime,omitempty"`
+	IoMerged       []PerDiskStats `json:"ioMerged,omitempty"`
+	IoTime         []PerDiskStats `json:"ioTime,omitempty"`
+}
+
+type MemoryStats struct {
+	// Current memory usage, this includes all memory regardless of when it was
+	// accessed.
+	// Units: Bytes.
+	Usage uint64 `json:"usage,omitempty"`
+
+	// The amount of working set memory, this includes recently accessed memory,
+	// dirty memory, and kernel memory. Working set is <= "usage".
+	// Units: Bytes.
+	WorkingSet uint64 `json:"workingSet,omitempty"`
+
+	// Cumulative number of times that a usage counter hit its limit
+	Failcnt uint64 `json:"failcnt,omitempty"`
+
+	ContainerData    MemoryStatsMemoryData `json:"containerData,omitempty"`
+	HierarchicalData MemoryStatsMemoryData `json:"hierarchicalData,omitempty"`
+}
+
+type MemoryStatsMemoryData struct {
+	// Cumulative number of minor page faults.
+	Pgfault uint64 `json:"pgfault,omitempty"`
```

Comment:
A lot of the field names in these structs are copied from the stat names reported by the linux kernel. On the one hand, it makes it clear what the fields are reporting. On the other hand, this could be viewed as an implementation detail which should be masked.